### PR TITLE
Add lab1 part1: learning switch

### PR DESCRIPTION
**Pull request by sding3**
Add lab1 part1: learning switch

Testing result using class provided unit tests:

```bash
root@bba6c20d710d:/switchyard/switchyard# ./swyard.py -t /tmp/myswitchfifo_test.srpy /lab1/myswitch_fifo.py -v
01:45:32 2019/10/10     INFO Starting test scenario
/tmp/myswitchfifo_test.srpy

Results for test scenario Packet forwrding in fifo example: 14 passed, 0
failed, 0 pending

Passed:
1   Ethernet frame from 20:00:00:00:00:01 to 30:00:00:00:00:02
    with unknown destination port should arrive on eth0
        Expected event: recv_packet Ethernet
        20:00:00:00:00:01->30:00:00:00:00:02 IP | IPv4
        192.168.1.100->192.168.1.101 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) on eth0
2   Packet to 30:00:00:00:00:02 should be flooded along eth1 and
    eth2
        Expected event: send_packet(s) Ethernet
        20:00:00:00:00:01->30:00:00:00:00:02 IP | IPv4
        192.168.1.100->192.168.1.101 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth1 and Ethernet
        20:00:00:00:00:01->30:00:00:00:00:02 IP | IPv4
        192.168.1.100->192.168.1.101 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth2
3   Ethernet frame from 30:00:00:00:00:02 to 40:00:00:00:00:03
    with unknown destination port should arrive on eth1
        Expected event: recv_packet Ethernet
        30:00:00:00:00:02->40:00:00:00:00:03 IP | IPv4
        192.168.1.101->192.168.1.102 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) on eth1
4   Packet to 30:00:00:00:00:02 should be flooded along eth0 and
    eth2
        Expected event: send_packet(s) Ethernet
        30:00:00:00:00:02->40:00:00:00:00:03 IP | IPv4
        192.168.1.101->192.168.1.102 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth0 and Ethernet
        30:00:00:00:00:02->40:00:00:00:00:03 IP | IPv4
        192.168.1.101->192.168.1.102 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth2
5   Ethernet frame from 40:00:00:00:00:03 to 50:00:00:00:00:04
    with unknown destination port should arrive on eth0
        Expected event: recv_packet Ethernet
        40:00:00:00:00:03->50:00:00:00:00:04 IP | IPv4
        192.168.1.102->192.168.1.103 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) on eth0
6   Packet to 30:00:00:00:00:02 should be flooded along eth1 and
    eth2
        Expected event: send_packet(s) Ethernet
        40:00:00:00:00:03->50:00:00:00:00:04 IP | IPv4
        192.168.1.102->192.168.1.103 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth1 and Ethernet
        40:00:00:00:00:03->50:00:00:00:00:04 IP | IPv4
        192.168.1.102->192.168.1.103 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth2
7   Ethernet frame from 20:00:00:00:00:01 to 30:00:00:00:00:02
    with known destination port should arrive on eth0
        Expected event: recv_packet Ethernet
        20:00:00:00:00:01->30:00:00:00:00:02 IP | IPv4
        192.168.1.100->192.168.1.101 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) on eth0
8   Packet to 30:00:00:00:00:02 should be flooded along eth1
    since the destination port should have been learnt before
        Expected event: send_packet(s) Ethernet
        20:00:00:00:00:01->30:00:00:00:00:02 IP | IPv4
        192.168.1.100->192.168.1.101 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth1
9   Ethernet frame from 5:00:00:00:00:04 to 60:00:00:00:00:05
    with unknown destination port should arrive on eth2
        Expected event: recv_packet Ethernet
        50:00:00:00:00:04->60:00:00:00:00:05 IP | IPv4
        192.168.1.103->192.168.1.104 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) on eth2
10  Packet to 60:00:00:00:00:05 should be flooded along eth0 and
    eth1
        Expected event: send_packet(s) Ethernet
        50:00:00:00:00:04->60:00:00:00:00:05 IP | IPv4
        192.168.1.103->192.168.1.104 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth0 and Ethernet
        50:00:00:00:00:04->60:00:00:00:00:05 IP | IPv4
        192.168.1.103->192.168.1.104 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth1
11  Ethernet frame from 60:00:00:00:00:04 to 70:00:00:00:00:06
    with unknown destination port should arrive on eth1
        Expected event: recv_packet Ethernet
        60:00:00:00:00:04->70:00:00:00:00:06 IP | IPv4
        192.168.1.104->192.168.1.105 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) on eth1
12  Packet to 70:00:00:00:00:06 should be flooded along eth0 and
    eth2
        Expected event: send_packet(s) Ethernet
        60:00:00:00:00:04->70:00:00:00:00:06 IP | IPv4
        192.168.1.104->192.168.1.105 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth0 and Ethernet
        60:00:00:00:00:04->70:00:00:00:00:06 IP | IPv4
        192.168.1.104->192.168.1.105 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth2
13  Ethernet frame from 20:00:00:00:00:01 to 30:00:00:00:00:02
    with unknown destination port should arrive on eth2
        Expected event: recv_packet Ethernet
        70:00:00:00:00:06->20:00:00:00:00:01 IP | IPv4
        192.168.1.105->192.168.1.100 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) on eth2
14  Packet to 30:00:00:00:00:02 should be flooded along eth0 and
    eth1 since entry for the destination 20:00:00:00:00:01
    should have been removed by FIFO policy
        Expected event: send_packet(s) Ethernet
        70:00:00:00:00:06->20:00:00:00:00:01 IP | IPv4
        192.168.1.105->192.168.1.100 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth0 and Ethernet
        70:00:00:00:00:06->20:00:00:00:00:01 IP | IPv4
        192.168.1.105->192.168.1.100 ICMP | ICMP EchoRequest 0 0 (0
        data bytes) out eth1

All tests passed!
```

### CODE REVIEW

**Comment by sding3** [comment]
@dmbillmireUW could you review? Once approved, I'll merge into master

File: lab1/myswitch_fifo.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
+from switchyard.lib.userlib import *
+
+
+class ForwaridngTable:
```
Comment (dmbillmireUW) [inline lab1/myswitch_fifo.py]: Should be "ForwardingTable" (same transpose in comment & when used in main()

File: lab1/myswitch_fifo.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            log_debug("Packet intended for me")
+        else:
+            if packet[0].dst in forwarding_table:
+                net.send_packet(forwarding_table[packet[0].dst], packet)
+            else:
+                for intf in my_interfaces:
+                    if input_port != intf.name:
+                        log_debug("Flooding packet {} to {}".format(
+                            packet, intf.name))
+                        net.send_packet(intf.name, packet)
```
Comment (dmbillmireUW) [inline lab1/myswitch_fifo.py]: It's unlikely, but for completeness shall we add a ValueError exception handler?

**Review by dmbillmireUW — CHANGES_REQUESTED** [review]
Looks good -- only a few nit picks then this is good to merge.
...
Except I just noticed we need to handle the broadcast MAC too.

**Review by dmbillmireUW — CHANGES_REQUESTED** [review]
We'll need to handle frames sent to the broadcast MAC of FF:FF:FF:FF:FF:FF

File: lab1/myswitch_fifo.py (hunk 2 above)
Comment (sding3) [reply lab1/myswitch_fifo.py]: Yes, we should handle that. I've pushed updates.

File: lab1/myswitch_fifo.py (hunk 1 above)
Comment (sding3) [reply lab1/myswitch_fifo.py]: Corrected now.

**Comment by sding3** [comment]
> We'll need to handle frames sent to the broadcast MAC of FF:FF:FF:FF:FF:FF

Good catch! I've added handling for it now.

**Review by dmbillmireUW — APPROVED** [review]

**Pull request merged**